cqc/pythonLib.py

```diff
@@ -175,1790 +176,2250 @@ def createXtraHeader(command, values):
+
+        - **Arguments**
+
+            :nofify:     Do we wish to be notified when done.
+            :block:         Do we want the qubit to be blocked
+        """
+        # check if qubit is active
+        self.check_active()
+
+        if self._cqc.pend_messages:
+            # print info
+            logging.debug("App {} pends message: 'Reset qubit with ID {}'".format(self._cqc.name, self._qID))
+
+            self._cqc.pending_messages.append([self, CQC_CMD_RESET, int(notify), int(block)])
+        else:
+            # print info
+            logging.debug("App {} tells CQC: 'Reset qubit with ID {}'".format(self._cqc.name, self._qID))
+
+            self._cqc.sendCommand(self._qID, CQC_CMD_RESET, notify=int(notify), block=int(block))
+            if notify and self.notify:
+                message = self._cqc.readMessage()
+                self._cqc.print_CQC_msg(message)
+
+    def release(self, notify=True, block=False):
+        """
+        Release the current qubit
+        :param notify: Do we wish to be notified when done
+        :param block: Do we want the qubit to be blocked
+        :return:
+        """
+        return self._cqc.release_qubits([self], notify=notify, block=block)
+
+    def getTime(self, block=True):
+        """
+        Returns the time information of the qubit.
+        If now INF_TIME message is received, None is returned.
+
+        - **Arguments**
+
+            :block:         Do we want the qubit to be blocked
+        """
+        # check if qubit is active
+        self.check_active()
+
+        # print info
+        logging.debug("App {} tells CQC: 'Return time-info of qubit with ID {}'".format(self._cqc.name, self._qID))
+
+        self._cqc.sendGetTime(self._qID, notify=0, block=int(block))
+
+        # Return time-stamp
+        message = self._cqc.readMessage()
+        try:
+            otherHdr = message[1]
+            return otherHdr.datetime
+        except AttributeError:
+            return None
+
+
+class CQCHandler(ABC):
+    """This class defines the things any CQCHandler must do.
+
+    It is to be subclassed by the various actual classes that handle CQC, such
+    as CQCConnection and CQCToFile.
+    
+    Essential things that a CQCHandler must do:
+    - sendCommand/sendCmdXtra
+    - construct commands
+    - decide if to pend or send immediatly
+    - return qubit IDs
+    - return messages if those are expected back
+    - pass entanglement information
+    - active qubits
+    - release qubits
+    - get time?
+
+    More general(?):
+    - construct commands
+    - send commands
+    - pend or not
+    - return qubit IDs
+    - pass back messages if those are expected
+    """
+
+    def __init__(self, pend_messages=False, notify=True):
+
+        # Decides if to send each command as it is received, or only 
+        # when flushed
+        self.pend_messages = pend_messages
+        self.pending_messages = []
+
+        # Set an app ID
+        self._appID = 6
```

Comment:
Why six? I think the functionality for increasing the appID could be moved to the abstract class

Reply:
It was just a number that made it easier to recognize in the printed out commands. There is no real reason for it.